silverback/cluster/types.py

```diff
@@ -42,25 +48,27 @@ class ClusterConfiguration(BaseModel):
     memory: Annotated[int, Field(ge=0, le=120)] = 0  # defaults to 512 MiB
     """Total memory per bot (in GB, 0 means '512 MiB')"""
 
-    # NOTE: Configure # of workers based on cpu & memory settings
+    # NOTE: # of workers configured based on cpu & memory settings
 
-    # Runner configuration (Bytes 3-5)
+    # Runner configuration (Bytes 3-4)
     networks: Annotated[int, Field(ge=1, le=20)] = 1
     """Maximum number of concurrent network runners"""
 
     bots: Annotated[int, Field(ge=1, le=250)] = 1
-    """Maximum number of concurrent bots running"""
+    """Maximum number of concurrent running bots"""
 
-    triggers: Annotated[int, Field(ge=50, le=1000, multiple_of=5)] = 50
-    """Maximum number of task triggers across all running bots"""
```

Comment:
Sorry for the delay on this feedback, but just through about it as I'm going through this stuff on the cluster side.

question: Are you sure about removing the triggers limit? From our discussion it sounded like this was going to be the primary limitation to Alchemy spend.

Reply:
The more I think about it, the more that I think that the best way to refactor our infrastructure long-term will involve piplining blocks (probably using some sort of multiple node aggregation to find the best new head) and then having the ARS fetch all the requested events directly from extended block data (which contains logs) present in the pipeline

The websockets are fairly unreliable, and having multiple connections at once can create weird partial failure modes, which I would argue is actually kind of worse than a total failure mode (since bot logic is not likely to be complicated enough to handle partial trigger losses)

I've also been told that this is in fact the best way to do things from a node perspective, and you can squeeze a lot of latency out of the system (latency is a prime concern among traders) this way too, if we optimize our infrastructure.

But that would probably come in a much larger refactor. For now, we can just do our best to control our costs by making sure we charge the ARS (e.g. number of networks) based on an average usage we are seeing. We also don't have to worry about dynamic changes to task triggers (yet, but it's on my personal roadmap), so we can just refuse to run a bot with a ridiculous among of triggers (maybe 100?) since we do query that via ARS startup sequence for a bot prior to putting it in the run state. That would encourage people to split up their bots with lots of triggers into multiple bots anyways, I would think that would probably be good architecturally too in order to maintain good bot performance and reduce complexity

Reply:
> The websockets are fairly unreliable, and having multiple connections at once can create weird partial failure modes, which I would argue is actually kind of worse than a total failure mode (since bot logic is not likely to be complicated enough to handle partial trigger losses)

*Our Websockets implementation* is unreliable and something we need to work on. We have no data to say that our provider's Websockets data streams are unreliable, however. This is something I hope to get more visibility into in the future. And if it becomes true, maybe something we can work out with them.

I'm not especially convinced fetching all block data for every block is the play either and may even increase costs for the majority of deployments. Might be in the cards for the future. Especially with a platform-level ARS if we end up needing to scale that much.

I'll put the hard non-ClusterConfiguration per-bot limit in for triggers on the docket though.

Reply:
> > The websockets are fairly unreliable, and having multiple connections at once can create weird partial failure modes, which I would argue is actually kind of worse than a total failure mode (since bot logic is not likely to be complicated enough to handle partial trigger losses)
> 
> *Our Websockets implementation* is unreliable and something we need to work on. We have no data to say that our provider's Websockets data streams are unreliable, however. This is something I hope to get more visibility into in the future. And if it becomes true, maybe something we can work out with them.

Yes our implementation has a few issues but also I have heard this from multiple devtool providers, node operators, traders, etc. all running a variety of different implementations and programming languages. No one is happy with how well the subscription websockets work. Also it is also going to get expensive if we have hundreds or thousands open all at the same time if they all have the same basic set of data (namely every cluster will most likely want newHead for each chain)

> I'm not especially convinced fetching all block data for every block is the play either and may even increase costs for the majority of deployments. Might be in the cards for the future. Especially with a platform-level ARS if we end up needing to scale that much.

You are definitely correct though that it is not something we need to be optimizing for now though, we need more data. This is my intuition though.